tensorboard/webapp/widgets/line_chart_v2/line_chart_component.ts

```diff
@@ -0,0 +1,348 @@
+          devicePixelRatio: window.devicePixelRatio,
+          callbacks,
+          domDimension: this.domDimensions.main,
+        };
+        break;
+      default:
+        const neverRendererType = rendererType as never;
+        throw new Error(
+          `<line-chart> does not yet support rendererType: ${neverRendererType}`
+        );
+    }
+
+    const useWorker =
+      rendererType !== RendererType.SVG && isOffscreenCanvasSupported();
+    const klass = useWorker ? WorkerChart : MainThreadChart;
+    this.lineChart = new klass(params);
+  }
+
+  ngOnDestroy() {
+    if (this.lineChart) this.lineChart.dispose();
+  }
+
+  getRendererType(): RendererType {
+    return getRendererType(this.preferredRendererType);
+  }
+
+  private readAndUpdateDomDimensions(): void {
+    this.domDimensions = {
+      main: {
+        width: this.main.nativeElement.clientWidth,
+        height: this.main.nativeElement.clientHeight,
+      },
+      xAxis: {
+        width: this.xAxis.nativeElement.clientWidth,
+        height: this.xAxis.nativeElement.clientHeight,
+      },
+      yAxis: {
+        width: this.yAxis.nativeElement.clientWidth,
+        height: this.yAxis.nativeElement.clientHeight,
+      },
+    };
+  }
+
+  /**
+   * Minimally and imperatively updates the chart library depending on prop changed.
+   */
+  private updateLineChart() {
+    if (!this.lineChart) return;
+
+    if (this.scaleUpdated) {
+      this.scaleUpdated = false;
+      this.lineChart.setXScaleType(this.xScaleType);
+      this.lineChart.setYScaleType(this.yScaleType);
+    }
+
+    if (this.isMetadataUpdated) {
+      this.isMetadataUpdated = false;
+      this.lineChart.setMetadata(this.seriesMetadataMap);
+    }
+
+    if (this.isDataUpdated) {
+      this.isDataUpdated = false;
+      this.lineChart.setData(this.seriesData);
+    }
+
+    // There are below conditions in which the viewExtent changes.
+    const viewBoxChange =
+      this.isFixedViewBoxUpdataed ||
+      this.isViewBoxUpdated ||
+      this.maybeSetViewBoxToDefault;
+
+    if (this.isFixedViewBoxUpdataed && this.fixedViewBox) {
+      this.viewBox = this.fixedViewBox;
+    } else if (this.maybeSetViewBoxToDefault) {
+      const dataExtent = computeDataSeriesExtent(
+        this.seriesData,
+        this.seriesMetadataMap
+      );
+
+      this.dataExtent = {
+        x: dataExtent.x ?? DEFAULT_EXTENT.x,
+        y: dataExtent.y ?? DEFAULT_EXTENT.y,
+      };
+      this.viewBox = this.getDefaultViewBox();
+    }
+
+    if (viewBoxChange) {
+      this.isFixedViewBoxUpdataed = false;
+      this.isViewBoxUpdated = false;
+      this.maybeSetViewBoxToDefault = false;
+      this.lineChart.setViewBox(this.viewBox);
+    }
+  }
+
+  onViewBoxChanged({dataExtent}: {dataExtent: Extent}) {
+    this.isViewBoxUpdated = true;
+    this.viewBox = dataExtent;
+    this.updateLineChart();
+  }
+
+  onViewBoxReset() {
+    this.maybeSetViewBoxToDefault = true;
```

Comment:
What I really like about the code right now is that `updateLineChart` efficiently makes only the necessary changes, and is written in a way such that we might be able to delay/throttle it if we wanted to.

However, some things I found confusing on a first read-through:
- `this.dataExtent` represents the extents of the time series data.  However, it's not updated whenever the series data changes.  Understanding when `this.dataExtent` is up-to-date vs. stale requires knowing how / when `maybeSetViewBoxToDefault` is set.
- If `this.dataExtent` can be stale, does that mean `getDefaultViewBox()`, which depends on it, can be stale as well?  Does this mean `this.viewBox` could be stale?  (as written, I think the PR ensures `this.viewBox` is never stale)
- We are reusing `this.maybeSetViewBoxToDefault` for 2 cases: [view reset on double click] and [view reset due to series data changing while the previous viewBox was equal to the default viewBox]

I'm curious what you think about this proposal:

Perhaps we can replace the `areExtentsEqual(defaultExtent, viewBox)` check entirely with a flag `isViewAtDefault`, and split out a flag `this.shouldResetToDefault` from `this.maybeSetViewBoxToDefault`

```ts
init() {
  this.isViewAtDefault = true;
}

ngOnChanges {
  if (changes['seriesData']) {
    // if view is at default, I care to compute.
    if (this.isViewAtDefault) {
      this.shouldResetToDefault = true;
    }
    // if view is not at default, no need to reset.
  }
}

updateLineChart() {
  if (this.shouldResetToDefault) {
    this.viewBox = this.computeDataSeriesExtent();
    this.shouldResetToDefault = false;
  }
}

onViewBoxChanged(newViewBox) {
  this.isViewAtDefault = false;
  this.shouldResetToDefault = false;
  this.viewBox = newViewBox;
  this.updateLineChart();
}

onViewBoxReset() {
  this.isViewAtDefault = true;
  this.shouldResetToDefault = true;
  this.updateLineChart();
}

```

Reply:
I incorporated some of it but not in verbatim.

Reply:
Thanks for incorporating this, I like the name `isViewBoxOverriden`.